File: src/assistant/rag/bm25.ts
/**
 * BM25 — In-memory Best Matching 25 index for KB chunk retrieval.
 *
 * US-912: Lexical ranking component of hybrid retrieval.
 * Standard BM25 with k1=1.2, b=0.75 parameters.
 * Supports multilingual queries (English, Malay, Chinese, Tamil).
 */

import type { KBChunk } from './chunker.js';

/** BM25 scored result */
export interface BM25Result {
  chunk: KBChunk;
  score: number;
  rank: number;
}

// BM25 tuning parameters
const K1 = 1.2;
const B = 0.75;

// Common stop words (English + Malay)
const STOP_WORDS = new Set([
  // English
  'a', 'an', 'the', 'is', 'are', 'was', 'were', 'be', 'been', 'being',
  'have', 'has', 'had', 'do', 'does', 'did', 'will', 'would', 'could',
  'should', 'may', 'might', 'shall', 'can', 'to', 'of', 'in', 'for',
  'on', 'with', 'at', 'by', 'from', 'as', 'into', 'about', 'between',
  'through', 'after', 'before', 'above', 'below', 'up', 'down', 'out',
  'off', 'over', 'under', 'and', 'but', 'or', 'nor', 'not', 'no',
  'so', 'if', 'then', 'than', 'that', 'this', 'it', 'its', 'i', 'me',
  'my', 'we', 'our', 'you', 'your', 'he', 'him', 'his', 'she', 'her',
  'they', 'them', 'their', 'what', 'which', 'who', 'when', 'where',
  'how', 'all', 'each', 'every', 'both', 'few', 'more', 'most', 'some',
  'any', 'such', 'only', 'very', 'just',
  // Malay
  'dan', 'atau', 'yang', 'di', 'ke', 'dari', 'untuk', 'dengan',
  'pada', 'ini', 'itu', 'adalah', 'akan', 'telah', 'sudah', 'juga',
  'tidak', 'bukan', 'ada', 'saya', 'kami', 'kita', 'anda', 'mereka',
]);

/**
 * Tokenize and normalize text for BM25.
 * Lowercases, removes punctuation, filters stop words.
 */
function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[^\w\s\u4e00-\u9fff\u0b80-\u0bff]/g, ' ') // Keep CJK + Tamil chars
    .split(/\s+/)
    .filter(t => t.length > 1 && !STOP_WORDS.has(t));
}

/**
 * In-memory BM25 index over KB chunks.
 */
export class BM25Index {
  private documents: KBChunk[] = [];
  private docTokens: string[][] = [];
  private docLengths: number[] = [];
  private avgDocLength = 0;
  private termDocFreq = new Map<string, number>(); // term → number of docs containing it
  private N = 0; // total number of documents

  /**
   * Build the BM25 index from KB chunks.
   */
  index(chunks: KBChunk[]): void {
    const start = Date.now();
    this.documents = chunks;
    this.N = chunks.length;
    this.docTokens = [];
    this.docLengths = [];
    this.termDocFreq = new Map();

    let totalLength = 0;

    for (const chunk of chunks) {
      const tokens = tokenize(chunk.text);
      this.docTokens.push(tokens);
      this.docLengths.push(tokens.length);
      totalLength += tokens.length;

      // Count unique terms per document for DF
      const uniqueTerms = new Set(tokens);
      for (const term of uniqueTerms) {
        this.termDocFreq.set(term, (this.termDocFreq.get(term) || 0) + 1);
      }
    }

    this.avgDocLength = this.N > 0 ? totalLength / this.N : 0;
    console.log(`[RAG:BM25] Indexed ${this.N} chunks (${this.termDocFreq.size} unique terms) in ${Date.now() - start}ms`);
  }

  /**
   * Search the index with a query.
   *
   * @param query - User query text
   * @param topK - Number of top results to return
   * @returns Scored and ranked results
   */
  search(query: string, topK: number = 20): BM25Result[] {
    if (this.N === 0) return [];

    const queryTerms = tokenize(query);
    if (queryTerms.length === 0) return [];

    const scores: number[] = new Array(this.N).fill(0);

    for (const term of queryTerms) {
      const df = this.termDocFreq.get(term) || 0;
      if (df === 0) continue;

      // IDF: log((N - df + 0.5) / (df + 0.5) + 1)
      const idf = Math.log((this.N - df + 0.5) / (df + 0.5) + 1);

      for (let i = 0; i < this.N; i++) {
        // Term frequency in document
        const tf = this.countTerm(this.docTokens[i], term);
        if (tf === 0) continue;

        // BM25 score component
        const docLen = this.docLengths[i];
        const numerator = tf * (K1 + 1);
        const denominator = tf + K1 * (1 - B + B * (docLen / this.avgDocLength));
        scores[i] += idf * (numerator / denominator);
      }
    }

    // Get top-K results
    const results: BM25Result[] = [];
    const indices = scores
      .map((score, idx) => ({ score, idx }))
      .filter(item => item.score > 0)
      .sort((a, b) => b.score - a.score)
      .slice(0, topK);

    for (let rank = 0; rank < indices.length; rank++) {
      const { score, idx } = indices[rank];
      results.push({
        chunk: this.documents[idx],
        score,
        rank: rank + 1,
      });
    }

    return results;
  }

  /**
   * Count occurrences of a term in a token array.
   */
  private countTerm(tokens: string[], term: string): number {
    let count = 0;
    for (const t of tokens) {
      if (t === term) count++;
    }
    return count;
  }

  /** Number of indexed documents */
  get size(): number {
    return this.N;
  }
}

File: go-core/internal/rag/bm25.go
// Package rag is the Go port of assistant/rag (bm25.ts + chunker.ts): a sparse
// BM25 retriever over the profile's knowledge-base markdown files. It grounds
// llm_reply with the most relevant KB chunks. (Dense/vector rerank can be layered
// on later via the ML sidecar's /embed endpoint.)
package rag

import (
	"math"
	"strings"
	"unicode"
)

// BM25 parameters (standard, matching bm25.ts).
const (
	k1 = 1.2
	b  = 0.75
)

// Chunk is a retrievable KB segment.
type Chunk struct {
	Source string // filename
	Text   string
}

// Result is a scored chunk.
type Result struct {
	Chunk Chunk
	Score float64
}

var stopWords = map[string]bool{
	"the": true, "a": true, "an": true, "and": true, "or": true, "but": true, "is": true,
	"are": true, "was": true, "were": true, "to": true, "of": true, "in": true, "on": true,
	"at": true, "for": true, "with": true, "as": true, "by": true, "it": true, "this": true,
	"that": true, "be": true, "do": true, "does": true, "i": true, "you": true, "we": true,
	"can": true, "will": true, "my": true, "your": true,
	// Malay common
	"yang": true, "dan": true, "di": true, "ke": true, "untuk": true, "ada": true, "ini": true,
	"itu": true, "saya": true, "anda": true,
}

// tokenize mirrors bm25.ts: lowercase, keep word chars + CJK + Tamil, drop
// 1-char tokens and stop words.
func tokenize(text string) []string {
	var sb strings.Builder
	for _, r := range strings.ToLower(text) {
		if r == '_' || unicode.IsLetter(r) || unicode.IsDigit(r) || unicode.Is(unicode.Han, r) || unicode.Is(unicode.Tamil, r) {
			sb.WriteRune(r)
		} else {
			sb.WriteRune(' ')
		}
	}
	var out []string
	for _, t := range strings.Fields(sb.String()) {
		if len([]rune(t)) > 1 && !stopWords[t] {
			out = append(out, t)
		}
	}
	return out
}

// Index is an in-memory BM25 index over KB chunks.
type Index struct {
	docs        []Chunk
	docTokens   [][]string
	docLengths  []int
	avgDocLen   float64
	termDocFreq map[string]int
	n           int
}

// NewIndex builds a BM25 index from chunks.
func NewIndex(chunks []Chunk) *Index {
	idx := &Index{docs: chunks, n: len(chunks), termDocFreq: map[string]int{}}
	total := 0
	for _, c := range chunks {
		toks := tokenize(c.Text)
		idx.docTokens = append(idx.docTokens, toks)
		idx.docLengths = append(idx.docLengths, len(toks))
		total += len(toks)
		seen := map[string]bool{}
		for _, t := range toks {
			if !seen[t] {
				seen[t] = true
				idx.termDocFreq[t]++
			}
		}
	}
	if idx.n > 0 {
		idx.avgDocLen = float64(total) / float64(idx.n)
	}
	return idx
}

// Len returns the number of indexed chunks.
func (idx *Index) Len() int { return idx.n }

// Search returns the top-K chunks for a query, scored by BM25.
func (idx *Index) Search(query string, topK int) []Result {
	if idx.n == 0 {
		return nil
	}
	terms := tokenize(query)
	if len(terms) == 0 {
		return nil
	}
	scores := make([]float64, idx.n)
	for _, term := range terms {
		df := idx.termDocFreq[term]
		if df == 0 {
			continue
		}
		idf := math.Log((float64(idx.n)-float64(df)+0.5)/(float64(df)+0.5) + 1)
		for i := 0; i < idx.n; i++ {
			tf := countTerm(idx.docTokens[i], term)
			if tf == 0 {
				continue
			}
			docLen := float64(idx.docLengths[i])
			num := float64(tf) * (k1 + 1)
			den := float64(tf) + k1*(1-b+b*(docLen/idx.avgDocLen))
			scores[i] += idf * (num / den)
		}
	}
	// Collect positive-scored, sort desc, take topK.
	type sc struct {
		i int
		s float64
	}
	var ranked []sc
	for i, s := range scores {
		if s > 0 {
			ranked = append(ranked, sc{i, s})
		}
	}
	// simple insertion sort by score desc (topK small)
	for i := 1; i < len(ranked); i++ {
		for j := i; j > 0 && ranked[j].s > ranked[j-1].s; j-- {
			ranked[j], ranked[j-1] = ranked[j-1], ranked[j]
		}
	}
	if topK > len(ranked) {
		topK = len(ranked)
	}
	out := make([]Result, 0, topK)
	for i := 0; i < topK; i++ {
		out = append(out, Result{Chunk: idx.docs[ranked[i].i], Score: ranked[i].s})
	}
	return out
}

func countTerm(tokens []string, term string) int {
	c := 0
	for _, t := range tokens {
		if t == term {
			c++
		}
	}
	return c
}
